### PR TITLE
[Merged by Bors] - feat: More `Finset.sup'` lemmas

### PR DESCRIPTION
**Pull request by YaelDillies**
Match https://github.com/leanprover-community/mathlib/pull/18989

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/Finset/Lattice.lean [hunk 1]
```diff
@@ -124,6 +118,12 @@ theorem le_sup {b : β} (hb : b ∈ s) : f b ≤ s.sup f :=
 theorem le_sup_of_le {b : β} (hb : b ∈ s) (h : a ≤ f b) : a ≤ s.sup f := h.trans <| le_sup hb
 #align finset.le_sup_of_le Finset.le_sup_of_le
 
+theorem sup_union [DecidableEq β] : (s₁ ∪ s₂).sup f = s₁.sup f ⊔ s₂.sup f :=
+  Finset.induction_on s₁
+    (by rw [empty_union, sup_empty, bot_sup_eq])
+    (fun a s _ ih => by rw [insert_union, sup_insert, sup_insert, ih, sup_assoc])
```
Comment (eric-wieser) [inline Mathlib/Data/Finset/Lattice.lean:124]: The mathport output doesn't agree here, its proof is shorter.

File: Mathlib/Data/Finset/Basic.lean [hunk 2]
```diff
@@ -799,14 +799,33 @@ theorem eq_empty_of_ssubset_singleton {s : Finset α} {x : α} (hs : s ⊂ {x})
...
-  (eq_singleton_or_nontrivial ha).imp_left <| Exists.intro a
+theorem nontrivial_iff_ne_singleton (ha : a ∈ s) : s.Nontrivial ↔ s ≠ {a} :=
+  ⟨Nontrivial.ne_singleton, (eq_singleton_or_nontrivial ha).resolve_left⟩
+#align finset.nontrivial_iff_ne_singleton Finset.nontrivial_iff_ne_singleton
+
+theorem Nonempty.exists_eq_singleton_or_nontrivial : s.Nonempty → (∃ a, s = {a}) ∨ s.Nontrivial :=
+  fun ⟨a, ha⟩ => (eq_singleton_or_nontrivial ha).imp_left <| Exists.intro a
 #align finset.nonempty.exists_eq_singleton_or_nontrivial Finset.Nonempty.exists_eq_singleton_or_nontrivial
 
 instance [Nonempty α] : Nontrivial (Finset α) :=
```
Comment (eric-wieser) [inline Mathlib/Data/Finset/Basic.lean:831]: Can you add the #align here? I agree there's no need to prime the name any more.

**Comment by eric-wieser** [comment]
LGTM, thanks!

bors d+

Don't forget the CI failure, on top of my comments.

File: Mathlib/Data/Finset/Lattice.lean (hunk 1 above)
Comment (YaelDillies) [reply Mathlib/Data/Finset/Lattice.lean:124]: Found an even shorter proof, at the expense of reordering other lemmas a bit.

**Comment by YaelDillies** [comment]
Let's try!

bors merge

**Pull request closed without merging**